File: frontend/src/types/assurance.ts
export type AssuranceState = 'REVIEWABLE' | 'HOLD' | 'UNAVAILABLE'
export type AssuranceCaseStatus = 'OPEN' | 'ASSESSED' | 'REVIEWED' | 'ARCHIVED'
export type AssuranceMatrixStatus = 'PASS' | 'WARNING' | 'FAIL' | 'MISSING'

export interface AssuranceCaseCreate {
  assigned_reviewer_id?: string | null
  title: string
  purpose: string
  subject_type: string
  subject_key: string
  context: Record<string, unknown>
  required_roles: string[]
  policy_key: string
}

export interface AssuranceCase {
  assigned_reviewer_id?: string | null
  id: string
  user_id: string
  title: string
  purpose: string
  subject_type: string
  subject_key: string
  context: Record<string, unknown>
  required_roles: string[]
  policy_key: string
  policy_version: string
  status: AssuranceCaseStatus
  created_at: string
  updated_at: string
}

export interface AssuranceEvidence {
  id: string
  case_id: string
  source_id: string
  source_key: string
  source_name: string
  source_url?: string | null
  license_name?: string | null
  /** Forward-compatible with the source catalog; older APIs may omit this field. */
  license_url?: string | null
  attribution_text?: string | null
  terms_note?: string | null
  authority_level?: string
  official?: boolean
  source_limitations?: Record<string, unknown>
  evidence_role: string
  required: boolean
  entity_type: string
  entity_key: string
  canonical_source_type: string
  observed_at: string
  fetched_at: string
  valid_until?: string | null
  freshness_state: string
  availability_state: string
  confidence?: number | null
  completeness?: number | null
  value_summary: Record<string, unknown>
  metadata: Record<string, unknown>
  integrity_checksum?: string | null
  created_at: string
}

export interface AssuranceFinding {
  id: string
  code: string
  severity: 'INFO' | 'WARNING' | 'ERROR'
  message: string
  evidence_record_ids: string[]
  details: Record<string, unknown>
  created_at: string
}

export interface AssuranceMatrixItem {
  role: string
  required: boolean
  status: AssuranceMatrixStatus
  record_ids: string[]
  source_keys: string[]
  reason_codes: string[]
}

export interface AssuranceMatrix {
  case_id: string
  snapshot_id: string
  state: AssuranceState
  roles: AssuranceMatrixItem[]
}

export interface AssuranceAssessment {
  id: string
  case_id: string
  sequence_no: number
  decision_state: AssuranceState
  policy_key: string
  policy_version: string
  matrix: AssuranceMatrix
  metrics: Record<string, unknown>
  bundle_checksum: string
  bundle_signature: string
  signing_key_id: string
  signing_algorithm: string
  assessed_at: string
  findings: AssuranceFinding[]
}

export interface AssuranceTimelineEvent {
  occurred_at: string
  event_type: string
  entity_id: string
  summary: string
  details: Record<string, unknown>
}

export interface AssuranceTimeline {
  case_id: string
  events: AssuranceTimelineEvent[]
}

export interface AssuranceReviewReceipt {
  id: string
  case_id: string
  snapshot_id: string
  reviewer_id: string
  reviewer_role: string
  outcome: 'ATTESTED' | 'RETURNED'
  statement: string
  snapshot_checksum: string
  receipt_checksum: string
  receipt_signature: string
  signing_key_id: string
  signing_algorithm: string
  reviewed_at: string
}

export interface AssuranceBundle {
  format_version: 'clearpath.assurance-bundle.v1'
  case: AssuranceCase
  assessment: AssuranceAssessment
  evidence: AssuranceEvidence[]
  review_receipts: AssuranceReviewReceipt[]
  review_receipt_manifests: Record<string, unknown>[]
  sealed_manifest: Record<string, unknown>
  evidence_envelopes: Record<string, unknown>[]
  source_catalog: Record<string, unknown>[]
  lineage: Record<string, unknown>[]
  limitations: string[]
}

export interface AssuranceVerification {
  case_id: string
  snapshot_id: string
  verified: boolean
  bundle_checksum_valid: boolean
  bundle_signature_valid: boolean
  evidence_integrity_valid: boolean
  invalid_evidence_record_ids: string[]
  invalid_review_receipt_ids: string[]
}

export interface AssurancePolicy {
  key: string
  version: string
  title: string
  description: string
  decision_states: string[]
  checks: Array<Record<string, unknown>>
  limitations: string[]
}

File: backend/app/schemas/assurance.py
"""API contracts for generic non-vital evidence assurance."""

from __future__ import annotations

from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, Field, field_validator, model_validator


AssuranceState = Literal["REVIEWABLE", "HOLD", "UNAVAILABLE"]
CaseStatus = Literal["OPEN", "ASSESSED", "REVIEWED", "ARCHIVED"]


def _aware(value: datetime, name: str) -> datetime:
    if value.tzinfo is None or value.utcoffset() is None:
        raise ValueError(f"{name} must include a timezone")
    return value


class AssuranceCaseCreate(BaseModel):
    model_config = {"str_strip_whitespace": True}
    title: str = Field(..., min_length=3, max_length=200)
    purpose: str = Field(..., min_length=10, max_length=4000)
    subject_type: str = Field(..., min_length=1, max_length=80)
    subject_key: str = Field(..., min_length=1, max_length=255)
    context: dict = Field(default_factory=dict)
    required_roles: list[str] = Field(..., min_length=1, max_length=64)
    policy_key: str = Field(default="GENERIC_PROVENANCE_V1", max_length=80)
    assigned_reviewer_id: str | None = Field(default=None, max_length=64)

    @field_validator("required_roles")
    @classmethod
    def normalize_roles(cls, value: list[str]) -> list[str]:
        roles = [item.strip().upper() for item in value if item.strip()]
        if not roles:
            raise ValueError("at least one non-empty evidence role is required")
        if len(set(roles)) != len(roles):
            raise ValueError("required_roles must be unique")
        return roles

    @field_validator("assigned_reviewer_id")
    @classmethod
    def normalize_reviewer_id(cls, value: str | None) -> str | None:
        if value is None:
            return None
        normalized = value.strip()
        return normalized or None


class AssuranceCaseResponse(BaseModel):
    id: UUID
    user_id: str
    assigned_reviewer_id: str | None
    title: str
    purpose: str
    subject_type: str
    subject_key: str
    context: dict = Field(validation_alias="context_json", serialization_alias="context")
    required_roles: list[str]
    policy_key: str
    policy_version: str
    status: CaseStatus
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class AssuranceEvidenceCreate(BaseModel):
    model_config = {"str_strip_whitespace": True}
    source_id: UUID
    evidence_role: str = Field(..., min_length=1, max_length=80)
    required: bool = True
    entity_type: str = Field(..., min_length=1, max_length=80)
    entity_key: str = Field(..., min_length=1, max_length=255)
    # This public endpoint captures human declarations or imported documents.
    # Provider and derived classifications are assigned only by trusted server
    # connectors/transforms and can be linked through AssuranceEvidenceLinkCreate.
    canonical_source_type: Literal["OPERATOR_INPUT", "IMPORTED_DOCUMENT"]
    observed_at: datetime
    fetched_at: datetime
    valid_until: datetime | None = None
    freshness_state: Literal["FRESH", "AGING", "STALE", "NOT_APPLICABLE"]
    availability_state: Literal["AVAILABLE", "PARTIAL", "UNAVAILABLE"]
    confidence: float | None = Field(default=None, ge=0, le=1)
    completeness: float = Field(..., ge=0, le=1)
    value_summary: dict
    metadata: dict = Field(default_factory=dict)
    transform_name: str | None = Field(default=None, max_length=120)
    transform_version: str | None = Field(default=None, max_length=40)
    formula_reference: str | None = Field(default=None, max_length=120)
    parent_record_ids: list[UUID] = Field(default_factory=list, max_length=128)

    @model_validator(mode="after")
    def validate_evidence(self) -> "AssuranceEvidenceCreate":
        self.evidence_role = self.evidence_role.strip().upper()
        _aware(self.observed_at, "observed_at")
        _aware(self.fetched_at, "fetched_at")
        if self.valid_until is not None:
            _aware(self.valid_until, "valid_until")
        if self.fetched_at < self.observed_at:
            raise ValueError("fetched_at cannot be before observed_at")
        if not self.value_summary:
            raise ValueError("value_summary must contain the captured evidence payload")
        if len(set(self.parent_record_ids)) != len(self.parent_record_ids):
            raise ValueError("parent_record_ids must be unique")
        return self


class AssuranceEvidenceLinkCreate(BaseModel):
    """Link an already-ingested, owner-scoped provenance record to a case."""

    record_id: UUID
    evidence_role: str = Field(..., min_length=1, max_length=80)
    required: bool = True

    @field_validator("evidence_role")
    @classmethod
    def normalize_role(cls, value: str) -> str:
        normalized = value.strip().upper()
        if not normalized:
            raise ValueError("evidence_role cannot be empty")
        return normalized


class AssuranceEvidenceResponse(BaseModel):
    id: UUID
    case_id: UUID
    source_id: UUID
    source_key: str
    source_name: str
    source_url: str | None
    license_name: str | None
    license_url: str | None
    attribution_text: str | None
    terms_note: str | None
    authority_level: str
    official: bool
    source_limitations: dict
    evidence_role: str
    required: bool
    entity_type: str
    entity_key: str
    canonical_source_type: str
    observed_at: datetime
    fetched_at: datetime
    valid_until: datetime | None
    freshness_state: str
    availability_state: str
    confidence: float | None
    completeness: float | None
    value_summary: dict
    metadata: dict
    integrity_checksum: str | None
    created_at: datetime


class AssuranceFindingResponse(BaseModel):
    id: UUID
    code: str
    severity: Literal["INFO", "WARNING", "ERROR"]
    message: str
    evidence_record_ids: list[str]
    details: dict = Field(validation_alias="details_json", serialization_alias="details")
    created_at: datetime

    model_config = {"from_attributes": True}


class EvidenceMatrixItem(BaseModel):
    role: str
    required: bool
    status: Literal["PASS", "WARNING", "FAIL", "MISSING"]
    record_ids: list[UUID] = Field(default_factory=list)
    source_keys: list[str] = Field(default_factory=list)
    reason_codes: list[str] = Field(default_factory=list)


class EvidenceMatrixResponse(BaseModel):
    case_id: UUID
    snapshot_id: UUID
    state: AssuranceState
    roles: list[EvidenceMatrixItem]


class AssuranceAssessmentResponse(BaseModel):
    id: UUID
    case_id: UUID
    sequence_no: int
    decision_state: AssuranceState
    policy_key: str
    policy_version: str
    matrix: EvidenceMatrixResponse
    metrics: dict
    bundle_checksum: str
    bundle_signature: str
    signing_key_id: str
    signing_algorithm: str
    assessed_at: datetime
    findings: list[AssuranceFindingResponse]


class ReviewReceiptCreate(BaseModel):
    model_config = {"str_strip_whitespace": True}
    snapshot_id: UUID
    outcome: Literal["ATTESTED", "RETURNED"]
    statement: str = Field(..., min_length=12, max_length=4000)


class ReviewReceiptResponse(BaseModel):
    id: UUID
    case_id: UUID
    snapshot_id: UUID
    reviewer_id: str
    reviewer_role: str
    outcome: Literal["ATTESTED", "RETURNED"]
    statement: str
    snapshot_checksum: str
    receipt_checksum: str
    receipt_signature: str
    signing_key_id: str
    signing_algorithm: str
    reviewed_at: datetime

    model_config = {"from_attributes": True}


class AssuranceTimelineEvent(BaseModel):
    occurred_at: datetime
    event_type: str
    entity_id: UUID
    summary: str
    details: dict = Field(default_factory=dict)


class AssuranceTimelineResponse(BaseModel):
    case_id: UUID
    events: list[AssuranceTimelineEvent]


class AssuranceBundleResponse(BaseModel):
    format_version: Literal["clearpath.assurance-bundle.v1"] = (
        "clearpath.assurance-bundle.v1"
    )
    case: AssuranceCaseResponse
    assessment: AssuranceAssessmentResponse
    evidence: list[AssuranceEvidenceResponse]
    review_receipts: list[ReviewReceiptResponse]
    review_receipt_manifests: list[dict]
    sealed_manifest: dict
    evidence_envelopes: list[dict]
    source_catalog: list[dict]
    lineage: list[dict]
    limitations: list[str]


class AssuranceVerifyRequest(BaseModel):
    snapshot_id: UUID | None = None


class AssuranceVerifyResponse(BaseModel):
    case_id: UUID
    snapshot_id: UUID
    verified: bool
    bundle_checksum_valid: bool
    bundle_signature_valid: bool
    evidence_integrity_valid: bool
    invalid_evidence_record_ids: list[UUID]
    invalid_review_receipt_ids: list[UUID]


class AssurancePolicyResponse(BaseModel):
    key: str
    version: str
    title: str
    description: str
    decision_states: list[str]
    checks: list[dict]
    limitations: list[str]
